src/algebra/big_operators/intervals.lean

```diff
@@ -164,4 +166,119 @@ by rw [← sum_range_id_mul_two n, nat.mul_div_cancel]; exact dec_trivial
 
 end gauss_sum
 
+end generic
+
+section nat
+
+variable {β : Type*}
+variables (f g : ℕ → β) {m n : ℕ}
+
+-- The partial sums (starting from `0`) of `f` and `g` respectively.
+local notation `F` n:80 := ∑ i in range n, f i
+local notation `G` n:80 := ∑ i in range n, g i
+
+section monoid
+variable [add_comm_monoid β]
+
+lemma sum_Ico_empty : (∑ i in Ico n n, f i) = 0 :=
+begin
+  convert sum_empty,
+  exact Ico_self n,
+end
+
+lemma sum_Ico_single : (∑ i in Ico n (n+1), f i) = f n :=
+by rw [sum_eq_sum_Ico_succ_bot (lt_add_one n), sum_Ico_empty, add_zero]
+
+lemma sum_Ico_sub {c: ℕ} (hm : c ≤ m) (hmn : m ≤ n) :
+  ∑ (i : ℕ) in Ico (m-c) (n-c), f (i+c) = ∑ (i : ℕ) in Ico m n, f i :=
+begin
+  have hn : c ≤ n := le_trans hm hmn,
+  conv in (f (_+c)) { rw add_comm },
+  rw [sum_Ico_add f (m-c) (n-c), nat.sub_add_cancel hm, nat.sub_add_cancel hn],
+end
+
+end monoid
+
+section group
+variable [add_comm_group β]
+
+lemma sum_Ico_succ_sub_sum_eq_top : F (n+1) - F n = f n :=
+sub_eq_iff_eq_add'.mpr $ sum_range_succ f n
+
+lemma sum_Ico_succ_sub_top_eq_sum : F (n+1) - f n = F n :=
+by rw [sub_eq_iff_eq_add, ← sub_eq_iff_eq_add', sum_Ico_succ_sub_sum_eq_top]
+
+lemma sum_Ico_succ_bot' (hmn : m < n) :
+  ∑ i in Ico m n, f i - f m = ∑ i in Ico (m + 1) n, f i :=
+begin
+  have hmn1 : (m+1) ≤ n := nat.succ_le_iff.mpr hmn,
+  rw [sub_eq_iff_eq_add', ← sum_Ico_single f, sum_Ico_consecutive _ m.le_succ hmn1],
+end
+
+lemma sum_Ico_succ_top' (hmn : m ≤ n) :
+  ∑ i in Ico m (n+1), f i - f n = ∑ i in Ico m n, f i :=
+by rw [sub_eq_iff_eq_add', add_comm (f n), ← sum_Ico_succ_top hmn]
+
+end group
+
+variable [comm_ring β]
+
+/-- **Summation by parts**, also known as **Abel's lemma** or an **Abel transformation** -/
+theorem sum_by_parts (hmn : m < n) :
+  ∑ i in Ico m n, (f i * g i) =
```

Comment:
Perhaps you can help me understand the culture of proof golfing that seems to be the standard in Mathlib?

I totally understand minimizing the number of "helper" lemmas that aren't broadly useful (like `sum_Ico_single` above, although it is useful, just duplicative with `sum_singleton`), and there are situations where a better technique can dramatically shorten a proof ([like the use of asymptotics in this thread](https://leanprover.zulipchat.com/login/#narrow/stream/113488-general/topic/.E2.9C.94.20Proving.20.60tendsto.60.20lemmas.20with.20filters)). However, I don't see the point of rewriting things just to minimize the number of lines.

The proof of this identity is fundamentally a series of tedious algebraic manipulations; I doubt there's a technique in Lean that would change that fact. I could shorten it by replacing the `calc` proof with a combination of `simp`, `rewrite`, `conv` and `nth_rewrite`-a Lean expert could certainly do even better-but the whole point of `calc` is to write proofs that are longer but easier to repair since each step is recorded explicitly. Admittedly this particular one is kind of unreadable due to the 100 character line limit. I don't find the argument about compilation-time compelling. This `calc` proof compiles pretty fast (it's just rewrites and `ring` after all). Meanwhile there are [hundreds of unexpanded invocations of `tidy`](https://leanprover.zulipchat.com/login/#narrow/stream/113488-general/topic/.60tidy.60.20in.20mathlib.20proofs) to finish proofs, some of which take several seconds apiece.

To be clear, these questions don't arise solely from your comment. I've written several Mathlib PRs over the last few weeks that have received suggestions for shortening proofs by combining rewrites, eliminating variables, or switching to term mode. Is golfing proofs skill that I'll need to develop if I want to continue contributing to Mathlib? Should I avoid `calc` in general, since it tends to result in longer proofs? Am I ignoring some constraints that makes these kind of transformations desirable?

Reply:
@j-loreaux, all those points make sense in the broader context of Mathlib. Searching for the most "elegant" proof of has been part of mathematics long before proof assistants were. I guess I'm trying to understand what @b-mehta is asking for specifically when they say:

>  I wonder if the proof can be simplified!

Of course it can! If term-mode proofs are considered "simpler" than tactic mode proofs, then *every* tactic mode proof can be simplified. Your second point is about aesthetics (readibility/simplicity/beauty/comprehensibility). Does this proof violate some aesthetic principles that I'm not aware of? Should I collapse a few `calc` lines to avoid showing tedious steps? Switch away from `calc` entirely? Is there a tactic that could be used here that I don't know about? I don't want to spend time simplifying something only to be told that I'm doing it the wrong way.

In your first point, you mention that 

> by trying to golf a proof, you gain a better understanding of what the key steps / transformations are

but I don't want to get a better understanding of how to prove summation by parts. Summation by parts is boring. I just want it to exist in the library so I can prove something slightly less boring.

> mathlib isn't a textbook for learning

I wrote the proof this way t because the fastest way for me to write it was to copy-paste each step, change it incrementally to match what I had on paper, and then prove that the incremental change was correct. If this way of doing things is not suitable for inclusion in Mathlib, then I will change how I do things. By understanding why golfing is important, I hope to understand what kind of proofs *are* suitable for inclusion, and write those instead.

Reply:
First, please note that I am not criticizing you, only providing information. I can't speak for @b-mehta, or even more generally for mathlib maintainers, but I far as I am concerned and according to my own experience:

> If term-mode proofs are considered "simpler" than tactic mode proofs

Not necessarily. A 3-line term proof that could be closed in tactic mode via `by simp` is not preferable to the `simp` proof. On the other hand, a long calc proof which is boring for humans (but shows all the type transformations) may not be preferable to a term proof.

> Should I collapse a few `calc` lines to avoid showing tedious steps?

Definitely, especially successive rewrites, but also maybe find ways to restructure so that you don't have to do so many calculations "by hand".

> Switch away from `calc` entirely?

Not necessarily here, and definitely don't always avoid it, but use whatever works best, which is obviously context dependent.

> I wrote the proof this way because I thought the fastest way for me to write it was to copy-paste each step, change it incrementally to match what I had on paper, and then prove that the incremental change was correct.

That can be a reasonable way to write a first proof, but then afterwards you want to clean it up as much as possible. By this I mean golfing, or maybe writing a few key steps with `have`s, or looking for bits in the library that you can apply. Actually, this last bit is one reason _not_ to go straight from paper to Lean, because the existing API may affect decisions regarding proof structure. Of course, in this particular case the API may be limited, I'm not sure.

I could be wrong, but I think Bhavik's comment may be related to my first reaction to seeing the proof: "wow, that seems long. There's probably a simpler and shorter way." But without delving into the details of the proof and trying to rewrite some of it myself, I'm not immediately sure where to simplify (except for trivial things like chaining rewrites).

Reply:
I am still unclear as to what exactly @b-mehta wanted here. I've golfed this proof to get it as short as I can make it. Hopefully this will no longer trigger the "wow, that seems long" reaction, although I didn't find the process particularly "fun" :smile:.

I'm not aware of anything in the library that would have helped, so I used `rewrite_search` and some carefully targeted `conv` to write a non-`calc` version. As a result, this is less repairable than my original, but that does not seem to be a priority, and the original still exists in the commit history of this PR in case it does break in the future.

Reply:
I think @j-loreaux clarified precisely what I meant very well, in particular that
"a better understanding of what the key steps / transformations are, and which are not so important. Sometimes you even find that steps you thought were important actually weren't". 
My intention with this comment was not to criticise you, nor your proof, and I am sorry that it came across that way! Instead, it was just an immediate reaction that I had somewhat expected a more direct proof for this statement: simpler in the sense of conceptual ideas in the proof rather than proof lines. As you correctly pointed out, term-mode proofs can be both harder to read and harder to maintain, but if there are many ideas in one proof then it can be desirable to take some of them out into their own statements: there's a good chance that a useful idea here might be useful to someone else. And this would of course result in a shorter proof for your lemma as well.
But overall, the new proof you have here does look very clean to me, so thanks for your efforts!